src/hotspot/share/nmt/mallocHeader.hpp

```diff
@@ -110,14 +112,47 @@ class MallocHeader {
 
   static uint16_t build_footer(uint8_t b1, uint8_t b2) { return (uint16_t)(((uint16_t)b1 << 8) | (uint16_t)b2); }
 
-  uint8_t* footer_address() const   { return ((address)this) + sizeof(MallocHeader) + _size; }
-  uint16_t get_footer() const       { return build_footer(footer_address()[0], footer_address()[1]); }
-  void set_footer(uint16_t v)       { footer_address()[0] = (uint8_t)(v >> 8); footer_address()[1] = (uint8_t)v; }
+  uint16_t get_footer() const {
+    AsanPoisoningHelper aph(reinterpret_cast<uint16_t*>(footer_address()));
+    return build_footer(footer_address()[0], footer_address()[1]);
+  }
+
+  void set_footer(uint16_t v) {
+    AsanPoisoningHelper aph(reinterpret_cast<uint16_t*>(footer_address()));
+    footer_address()[0] = (uint8_t)(v >> 8); footer_address()[1] = (uint8_t)v;
+  }
 
   template<typename InTypeParam, typename OutTypeParam>
   inline static OutTypeParam resolve_checked_impl(InTypeParam memblock);
 
+  void asan_poison_self() {
+      AsanPoisoningHelper<uint16_t>::poison_memory(&_canary);
+      AsanPoisoningHelper<uint16_t>::poison_memory(reinterpret_cast<uint16_t*>(footer_address()));
+      AsanPoisoningHelper<size_t>::poison_memory(&_size);
+      NOT_LP64(AsanPoisoningHelper<uint32_t>::poison_memory(&_alt_canary));
+  }
+
+  void asan_unpoison_self() {
+      AsanPoisoningHelper<uint16_t>::unpoison_memory(&_canary);
+      AsanPoisoningHelper<uint16_t>::unpoison_memory(reinterpret_cast<uint16_t*>(footer_address()));
+      AsanPoisoningHelper<size_t>::unpoison_memory(&_size);
+      NOT_LP64(AsanPoisoningHelper<uint32_t>::unpoison_memory(&_alt_canary));
+  }
+
 public:
+  #ifndef _LP64
+  inline uint32_t alt_canary() const {
+    AsanPoisoningHelper aph(&_alt_canary);
+    return _alt_canary;
+  }
+  inline void set_alt_canary(uint32_t value) {
+    AsanPoisoningHelper aph(&_alt_canary);
+      _alt_canary = value;
+  }
+  #endif
+
+  uint8_t* footer_address() const { return ((address)this) + sizeof(MallocHeader) + size(); }
+
```

Comment:
Why is this moved to the public interface? If it's for the tests, then use a fixture class and a `friend class` declaration in `MallocHeader`.